### DISCUSSION TITLE
Is there a clean way to always run all validators?

### DISCUSSION THREAD

**Discussion by moenoel** [pydantic/pydantic#3993] [Question] [answered]
Is there a clean way to always run all validators, even if one of them previously failed? For instance, with the following example model:

```python
from pydantic import BaseModel, PositiveInt, validator

class Foo(BaseModel):
    bar: PositiveInt

    @validator('bar')
    def vfoo1(cls, v: int) -> int:
        if v < 10:
            raise ValueError('Must be at least 10')
        return v

    @validator('bar')
    def vfoo2(cls, v: int) -> int:
        if v % 5 != 0:
            raise ValueError('Must be a multiple of 5')
        return v
```

`Foo(bar=-1)` will fail the `PositiveInt` valdation and `vfoo1()` and `vfoo2()` are never called.

`Foo(bar=6)` will fail at `vfoo1()` and `vfoo2()` is never called.

I'd like to always get all the errors that would come up, including with nested models.

**Comment by RamiAwar** [comment, marked as answer]
Doesn't look like it from the source code. Validators are looped over and when an error is encountered its raised directly.
from validate_model in main.py#978:
```
for validator in model.__pre_root_validators__:
        try:
            input_data = validator(cls_, input_data)
        except (ValueError, TypeError, AssertionError) as exc:
            return {}, set(), ValidationError([ErrorWrapper(exc, loc=ROOT_KEY)], cls_)
```

**Reply by moenoel** [reply]
Yeah, working myself back from there, there doesn't seem to be a sane way to do what I want without rewriting the whole thing.

Thanks for giving me the starting point for the nitty gritty of it.

**Reply by RamiAwar** [reply]
You can always write your own!
Something along the lines of:

```
errors = []
for validator in model.__pre_root_validators__:
        try:
            input_data = validator(cls_, input_data)
        except (ValueError, TypeError, AssertionError) as exc:
            errors.append(ValidationError([ErrorWrapper(exc, loc=ROOT_KEY)], cls_))
return errors
```

**Reply by moenoel** [reply]
The problem is that it doesn't stop there. `validate_model()`  calls the `validate()` method of each field which down the line also does a loop like the one you quote and so on and so forth. I'd rather not maintain my own fork of pydantic or monkey patch half the deep guts of it.

My initial approach was to do one big validator that collects a list of validation problems and throws a custom exception with that list that'll then be automatically included in the `ctx` field of the resulting `ValidationError`, but that got real ugly real fast and it's more working around pydantic than with it. It also doesn't cover type mismatch (unless I use a root_validator with `pre=True` and do my own type validation, I guess, but at that point why even use pydantic at all?).

So it seems I'll either have to accept that dealing with validation errors might take multiple rounds of fixing and looking for more errors or not using pydantic at all and rolling my own solution.

**Reply by RamiAwar** [reply]
I'm not suggesting modifying pydantic, just creating your own function validate_model(model) that would return a list of errors.

**Reply by moenoel** [reply]
That comes down to the same thing, though. If I don't modify or monkey patch pydantic, I'd now somehow have to jerry rig the model to not do validation automatically and then write my own code to rumage around the guts of the model to get it's validators (they are not all in `model.__pre_root_validators__` btw.) and call them myself.

Why even use pydantic at that point when I'd end up with way less complexity doing it all manually?
